### PR TITLE
posix: env: support for environ, getenv(), setenv(), unsetenv()

### LINKED ISSUE #66861: posix: implement `environ`
This is part of the See [POSIX Roadmap for LTSv3](https://github.com/zephyrproject-rtos/zephyr/issues/51211) (RFC #51211).

The `environ` global variable is required as part of `POSIX_SINGLE_PROCESS` [Subprofiling Option Group](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html).

https://pubs.opengroup.org/onlinepubs/9699919799/functions/environ.html

### LINKED ISSUE #66862: posix: implmement `getenv()`
This is part of the See [POSIX Roadmap for LTSv3](https://github.com/zephyrproject-rtos/zephyr/issues/51211) (RFC #51211).

`getenv()` is required as part of `POSIX_SINGLE_PROCESS` [Subprofiling Option Group](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html).

For more information, please refer to https://pubs.opengroup.org/onlinepubs/9699919799/functions/getenv.html

### LINKED ISSUE #66863: posix: implement `setenv()`
This is part of the See [POSIX Roadmap for LTSv3](https://github.com/zephyrproject-rtos/zephyr/issues/51211) (RFC #51211).

`setenv()` is required as part of `POSIX_SINGLE_PROCESS` [Subprofiling Option Group](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html).

For more information, please refer to https://pubs.opengroup.org/onlinepubs/9699919799/functions/setenv.html

### LINKED ISSUE #66864: posix: implement `unsetenv()`
This is part of the See [POSIX Roadmap for LTSv3](https://github.com/zephyrproject-rtos/zephyr/issues/51211) (RFC #51211).

`unsetenv()` is required as part of `POSIX_SINGLE_PROCESS` [Subprofiling Option Group](https://pubs.opengroup.org/onlinepubs/9699919799/xrat/V4_subprofiles.html).

For more information, please refer to https://pubs.opengroup.org/onlinepubs/9699919799/functions/unsetenv.html

### PR DESCRIPTION
**Pull request by cfriedt**
**Note: Coding Guidelines failures are false positives**

Support getting and setting POSIX environment variables. Additionally, the thread-safe BSD variant `getenv_r()` is provided.

`environ`, `getenv()`, `setenv()`, and `unsetenv()` are required by the `POSIX_SINGLE_PROCESS` Option Group as detailed in Section E.1 of IEEE-1003.1-2017.

The `POSIX_SINGLE_PROCESS` Option Group is required for PSE51, PSE52, PSE53, and PSE54 conformance, and is otherwise mandatory for any POSIX conforming system as per Section A.2.1.3 of IEEE-1003-1.2017.

This lets us do fun things like what is shown below in the Zephyr shell:
```shell
uart:~$ posix env 
  set    get    unset
uart:~$ posix env get 
uart:~$ posix env set FOO=BAR
uart:~$ posix env get
FOO=BAR
uart:~$ posix env get FOO
BAR
uart:~$ posix env unset FOO
uart:~$ posix env set " 42 BCD" "No Problem!"
uart:~$ posix env get
uart:~$ posix env set "_BcD" "No Problem!"
uart:~$ posix env get
_BcD=No Problem!
uart:~$ posix env unset _BcD
```

App and library code can then query values from the environment, as a relatively simple and portable way of steering the behaviour of code at runtime.

Some notes:

* "poorly conditioned" variable names are dropped silently in the shell, which matches bash (or at least zsh)
* the Zephyr shell does not perform shell variable expansion at this time; in other words `echo "$VALUE"` will not work as one might expect

Possible future improvements:
* Heh... half joking, but adding shell variable expansion (or even an actual POSIX shell)
* this implementation is optimized for size (a speed optimization could maybe use hash tables)
* another future optimization for safety would be to have a malloc-free implementation
* a possibly nice feature for the shell would be to enable tab completions of names with `env get <tab>`
* another future improvement would be to use Kconfig / iterable sections to optionally enable some default environment variables to be populated (e.g. `HOME`, `PWD`, ...).
* another future improvement might be to add hooks for "dynamic" environment variables like `$RANDOM`.

Fixes #66861
Fixes #66862
Fixes #66863
Fixes #66864

### CODE REVIEW

File: subsys/shell/modules/env_service.c
Comment (cfriedt) [inline subsys/shell/modules/env_service.c:79]: There probably needs to be some additional checks to verify that names are "well conditioned".

E.g. `[_A-Z][_A-Z0-9]+`

**Comment by cfriedt** [comment]
Screenshot 2023-12-21 at 1 41 35 PM

File: tests/posix/env/testcase.yaml
Comment (cfriedt) [inline tests/posix/env/testcase.yaml:28]: These are excluded, because it was the 1 test case that was not passing.

There is an error about `_POSIX_C_SOURCE` being redefined, likely here
https://github.com/zephyrproject-rtos/picolibc/blob/1a5c603b9f8e228f9459bdafedad15ea28efc700/zephyr/zephyr.cmake#L95

```shell
[927/1025] Building C object modules/picolibc/CMakeFiles/c.dir/newlib/libm/complex/clogl.c.obj
[928/1025] Building C object modules/picolibc/CMakeFiles/c.dir/newlib/libm/complex/cpowl.c.obj
[929/1025] Building C object modules/picolibc/CMakeFiles/c.dir/newlib/libm/complex/csinhl.c.obj
[930/1025] Building C object modules/picolibc/CMakeFiles/c.dir/newlib/libm/complex/csinl.c.obj
[931/1025] Building C object modules/picolibc/CMakeFiles/c.dir/newlib/libm/complex/ctanl.c.obj
[932/1025] Building C object modules/picolibc/CMakeFiles/c.dir/newlib/libm/complex/ctanhl.c.obj
[933/1025] Linking C static library modules/picolibc/libc.a
[934/1025] Generating include/generated/syscall_dispatch.c, include/generated/syscall_list.h
[935/1025] Building C object zephyr/CMakeFiles/offsets.dir/arch/posix/core/offsets/offsets.c.obj
FAILED: zephyr/CMakeFiles/offsets.dir/arch/posix/core/offsets/offsets.c.obj 
ccache /usr/lib/ccache/gcc -DKERNEL -DK_HEAP_MEM_POOL_SIZE=0 -DTC_RUNID=6b8855be30e94afb6f819ebb6f871902 -D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 -D_XOPEN_SOURCE_EXTENDED -D__LINUX_ERRNO_EXTENSIONS__ -D__ZEPHYR__=1 -I/home/cfriedt/zephyrproject/zephyr/kernel/include -I/home/cfriedt/zephyrproject/zephyr/arch/posix/include -I/home/cfriedt/zephyrproject/zephyr/include -I/home/cfriedt/zephyrproject/zephyr/twister-out/native_sim/tests/posix/env/portability.posix.env.picolibc/zephyr/include/generated -I/home/cfriedt/zephyrproject/zephyr/soc/posix/inf_clock -I/home/cfriedt/zephyrproject/zephyr/boards/posix/native_sim -I/home/cfriedt/zephyrproject/zephyr/scripts/native_simulator/common/src/include -I/home/cfriedt/zephyrproject/zephyr/scripts/native_simulator/native/src/include -I/home/cfriedt/zephyrproject/zephyr/subsys/testsuite/include -I/home/cfriedt/zephyrproject/zephyr/subsys/testsuite/ztest/include -isystem /home/cfriedt/zephyrproject/zephyr/lib/libc/common/include -isystem /home/cfriedt/zephyrproject/zephyr/twister-out/native_sim/tests/posix/env/portability.posix.env.picolibc/modules/picolibc/picolibc/include -fno-strict-aliasing -Werror -Os -imacros /home/cfriedt/zephyrproject/zephyr/twister-out/native_sim/tests/posix/env/portability.posix.env.picolibc/zephyr/include/generated/autoconf.h -fno-printf-return-value -fno-common -g -gdwarf-4 -fdiagnostics-color=always -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/home/cfriedt/zephyrproject/zephyr/tests/posix/env=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/cfriedt/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/cfriedt/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -m32 -msse2 -mfpmath=sse -fvisibility=hidden -nostdinc -isystem /usr/lib/gcc/x86_64-linux-gnu/11/include -ffreestanding -fno-builtin -D_POSIX_C_SOURCE=200809 -std=c11 -MD -MT zephyr/CMakeFiles/offsets.dir/arch/posix/core/offsets/offsets.c.obj -MF zephyr/CMakeFiles/offsets.dir/arch/posix/core/offsets/offsets.c.obj.d -o zephyr/CMakeFiles/offsets.dir/arch/posix/core/offsets/offsets.c.obj -c /home/cfriedt/zephyrproject/zephyr/arch/posix/core/offsets/offsets.c
<command-line>: error: "_POSIX_C_SOURCE" redefined [-Werror]
<command-line>: note: this is the location of the previous definition
cc1: all warnings being treated as errors
ninja: build stopped: subcommand failed.
```

**Comment by aescolar** [comment]
> Oh - more fun! @aescolar, if you add the following to `samples/hello_world/prj.conf` and then run this under `native_sim_64`, `env get` will list all of the host environment variables too. Nifty! (output below sanitized)

...
Correction: *will list all of the host environment variables ~too~*
With that configuration (w CONFIG_EXTERNAL_LIBC) you are not building [lib/posix/env.c](https://github.com/zephyrproject-rtos/zephyr/pull/66762/files#diff-f1fc40083c47df2c40e5f67c4aecc1a1a226cb39ad1cb0da9ddf1911d0115696) at all, and are instead calling into the host libC accessing and modifying the host environment. Which is very unlikely what users would want.

Since the library is excluded in this case: https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/CMakeLists.txt#L7 
It would also make sense that all related kconfig's would be if'ed by the same condition to avoid confusing developers.

File: tests/posix/env/testcase.yaml
Comment (aescolar) [reply tests/posix/env/testcase.yaml:28]: This is due to building PICOLIBC from source, you can trigger it by building from source for any target:
```
cmake -DBOARD=qemu_x86 ../tests/posix/env/ -GNinja -DCONFIG_PICOLIBC=y -DCONFIG_PICOLIBC_USE_MODULE=y
```
So I would not blacklist native_sim, but address the issue instead.

Despite of that warning, the test executes and passes fine.

**Review by aescolar — DISMISSED** [review]

File: tests/posix/env/testcase.yaml
Comment (cfriedt) [reply tests/posix/env/testcase.yaml:28]: > So I would not blacklist native_sim, but address the issue instead.

Already underway. https://github.com/picolibc/picolibc/issues/649

> Despite of that warning, the test executes and passes fine.

Of course. The warning is escalated to an error though in CI, which is problematic.

Of course as well, it should really only be set when posix functionality is activated via Kconfig which is maybe more problematic.

**Comment by aescolar** [comment]
> That's a given, yes. I just found it funny that it was pulling in the host environment.

@cfriedt you tend to edit other peoples answers to answer yourself on top? :p

File: arch/posix/CMakeLists.txt
Comment (cfriedt) [reply arch/posix/CMakeLists.txt:136]: > There is a misunderstanding in this commit.

No. There is not.

You _think_ there is a misunderstanding (there is a difference).

> This was not added here due to the POSIX compatibility API, but to support the POSIX arch (and drivers) calls into the host C library (This is

Correct. That is understood.

> needed for NATIVE_APPLICATION (old ARCH_POSIX targets) and the drivers which are not yet ported to work with all libCs; And both of these are not compatible with CONFIG_POSIX_API).

That was my assumption.

So likely this whole CMake file should not be  pulled-in at all and a filter is missing from the testcase.yaml the naming and reatrictions of native sim + native posix are unfortunately not less confusing than the original native posix target.

In any case though, these options should eventually be removed once the drivers in question are ported. I assume native posix targets will just be dropped at that point?

> So it cannot just be removed, and replaced with it being added if CONFIG_POSIX_API.

I think there needs to be a filter added.

**Comment by cfriedt** [comment]
> > That's a given, yes. I just found it funny that it was pulling in the host environment.
> 
> @cfriedt you tend to edit other peoples answers to answer yourself on top? :p

Maybe 🤷🏼‍♂️

Sometimes fingers hit the wrong buttons on mobile while mobile.

**Comment by cfriedt** [comment]
Screenshot 2023-12-23 at 11 57 29 AM

File: arch/posix/CMakeLists.txt
Comment (aescolar) [inline arch/posix/CMakeLists.txt:136]: This comment is not correct. Please remove it.

**Review by aescolar — DISMISSED** [review]

File: arch/posix/CMakeLists.txt
Comment (cfriedt) [reply arch/posix/CMakeLists.txt:136]: Please explain how you feel it is not correct

Comment (cfriedt) [reply arch/posix/CMakeLists.txt:136]: Nevermind. I'll remove it and file a bug

File: arch/posix/CMakeLists.txt
Comment (cfriedt) [reply arch/posix/CMakeLists.txt:136]: > However, additionally, I think it is probably not correct to blanket-define this at an architecture level as part of the zephyr build.

Resolving this now, as it's been fixed by having a common Kconfig `CONFIG_NEED_POSIX_C_SOURCE`.

File: soc/xtensa/espressif_esp32/esp32s3/Kconfig.series
Comment (cfriedt) [inline soc/xtensa/espressif_esp32/esp32s3/Kconfig.series:15]: Probably should be reduced in scope to when the Wifi driver is selected. @sylvioalves - I'm curious about how this is actually being used in the esp32 wifi driver (hal?).

File: include/zephyr/sys/util.h
Comment (jukkar) [inline include/zephyr/sys/util.h:211]: Hmm, why `++(idx)` instead of more typical `(idx)++`, don't we miss now one value from the loop?

File: include/zephyr/sys/util.h
Comment (jukkar) [inline include/zephyr/sys/util.h:220]: ditto

File: include/zephyr/sys/util.h
Comment (cfriedt) [reply include/zephyr/sys/util.h:211]: Hehe.. all values are accounted for.

It's mainly to be more specific about  avoid the possibility of a compiler generating less efficient code. gcc and llvm are usually good about optimizing-out any temporaries if post-incrementing is used and no assignment is done, but some other compilers might not do that as well. With pre-incrementing, there is no temporary.

E.g. `a = b++` and `a = ++b` vs `b++` and `++b`

File: include/zephyr/sys/util.h
Comment (cfriedt) [reply include/zephyr/sys/util.h:220]: same

File: include/zephyr/sys/util.h
Comment (jukkar) [reply include/zephyr/sys/util.h:211]: Indeed :disappointed: I wonder what I was thinking here...

File: lib/libc/minimal/include/stdlib.h
Comment (evgeniy-paltsev) [inline lib/libc/minimal/include/stdlib.h:73]: Just wondering what this magic `(_POSIX_C_SOURCE - 0)` do and if it's possible to drop this zero subtraction?

Comment (cfriedt) [reply lib/libc/minimal/include/stdlib.h:73]: This is a pattern I've seen used in Picolibc and I guess Newlib.

I would be fine using the regular `L` prefixed version as well.

File: include/zephyr/sys/util.h
Comment (cfriedt) [reply include/zephyr/sys/util.h:211]: This is already in main actually, so I should rebase..

File: lib/libc/minimal/include/stdlib.h
Comment (cfriedt) [reply lib/libc/minimal/include/stdlib.h]: @evgeniy-paltsev - fixed!

File: tests/posix/env/src/env.c
Comment (cfriedt) [inline tests/posix/env/src/env.c:109]: Closed now (thanks @keith-packard 😁) but it might take a bit of time before the module is updated in `west.yml`

**Review by keith-packard — COMMENTED** [review]
Seems like these functions should live in libc/common as a POSIX C library will already provide implementations -- these functions aren't OS specific.

**Comment by cfriedt** [comment]
> Seems like these functions should live in libc/common

I would say that statement is maximally 25% true; `getenv()` is part of ISO C but `environ`, `setenv()`, and `unsetenv()` are not. It would not be possible for an external ISO C library to use those POSIX features with Zephyr.

I do agree that ISO C functions should go into `libc/common`, but a **significant** effort was made to ensure that external C libraries **do not** have access to `libc/common` as a part of #67048.

"Not naming names" (to quote from last week's TSC meeting), but many reviewers / release managers / maintainers seemed very quick to submit and / or approve of that change. I felt  quite a lot of undue pressure to follow suit and merge it 🤷. I would happily revert it though.

```cmake
if(NOT CONFIG_EXTERNAL_LIBC)
add_subdirectory(libc)
endif()
```

> as a POSIX C library will already provide implementations -- these functions aren't OS specific.

Yes.

Picolibc does provide implementations of all of these. If `CONFIG_PICOLIBC=y`, and if the any of these symbols are not already pulled in by the time the libc is linked, Picolibc's version will be used. I believe the same already goes for Newlib.

Some issues preventing us from using other versions as of today:
* Picolibc currently doese not define `_POSIX_VERSION` when built against Zephyr, even if `_POSIX_C_SOURCE` is defined
* there is currently a bug in newlib's version of `setenv()`, so using the `lib/posix` version is probably a safer bet
* there was a bug in Picolibc's verison of `setenv()` and it will fail tests until `west.yml` and the SDK are updated. For the time being, the `lib/posix` version is probably a safer bet.

For those reasons, and because I agree that POSIX C libraries should very much provide those symbols, I only used `imply POSIX_ENV` for shell env service (rather than `select`). Otherwise, `CONFIG_POSIX_ENV` could be set to `n` and the shell env service should work fine (assuming `_POSIX_VERSION` is defined by the implementation).

Aside from `getenv()`, the remainder of these functions have nothing to do with ISO C and probably *should* go into `lib/posix`. The main reasons why I thought it useful to include `getenv()` in the POSIX implementation were

* there was no previous implementation of `getenv()` in the Minimal C library
* there is no requirement for the Minimal C library to have `getenv()`
* it is important that `getenv()`, `setenv()`, `unsetenv()` and `environ` accesses all be synchronized to avoid race conditions (so they should be co-located)
* `getenv()`, `setenv()`, `unsetenv()`, and `environ` are all requirements for PSE51, PSE52, and PSE53 (#51211)
* the commit(s) that fixed Picolibc and updated `west.yml` were in the future (at the time this change was made)

I would be happy to revisit at some point when the other issues are resolved.

Does that work for you @keith-packard ?

**Comment by cfriedt** [comment]
There is 1 error with `hifive1` when this is being simulated under Renode currently.
```shell
ASSERTION FAIL [((((size_t)16) << 10) - (((uintptr_t) (&_end)) - 0x80000000)) >= 8192] @ WEST_TOPDIR/zephyr/lib/libc/newlib/libc-hooks.c:128
19:20:58.0553 [INFO] uart0: [host: 0.16s (+0.47ms)|virt:  9µs (+0s)] memory space available for newlib heap is less than the minimum required size specified by CONFIG_NEWLIB_LIBC_MIN_REQUIRED_HEAP_SIZE
```

I just checked and it's set to 8192 for this build, which is well beyond the 256 bytes that the test requires. So I think that the issue is likely not related or it can likely be resolved via user configuration.

Tests are all passing on Tier 0 platforms otherwise, so I think I'm going to exclude that platform from the test.

**Comment by cfriedt** [comment]
Lots of false positives with "SonarCloud Code Analysis". Today is the first I've ever heard of it.

**Comment by cfriedt** [comment]
Corrected a few typos - CI is "green" again

**Comment by cfriedt** [comment]
@evgeniy-paltsev - Look ok for you for approval?
@ycsin ?

**Review by ycsin — DISMISSED** [review]
Implementation LGTM

File: tests/posix/env/testcase.yaml
Comment (aescolar) [inline tests/posix/env/testcase.yaml:2]: Minor clarity proposal: you can filter by NATIVE_LIBC instead

File: tests/posix/env/CMakeLists.txt
Comment (aescolar) [inline tests/posix/env/CMakeLists.txt:15]: Given that this gives trouble until the picolibc fix is merged, and given that this is just to work around checkpatch, I would either wait for picolibc's fix, or have it defined in the source, and ask for an exception as in this case it is perfectly justified as you want to test setenv() and unsetenv()

File: tests/posix/env/testcase.yaml
Comment (aescolar) [inline tests/posix/env/testcase.yaml:4]: Instead of excluding platforms that build picolibc from source, it would be better to undefine & define the macro in the .c file.
(Otherwise it is not just native_sim when building with picolibc but anybody building picolibc from source until keith's pr is merged).

File: subsys/shell/modules/CMakeLists.txt
Comment (aescolar) [inline subsys/shell/modules/CMakeLists.txt:23]: Same comment about having this in the source

File: tests/posix/env/src/env.c
Comment (aescolar) [inline tests/posix/env/src/env.c:77]: Nit: Using macros to replace trivial constructs is bad for readability.
Everybody who knows C can understand at a glance what
for (int i = 0; i < ARRAY_SIZE(args) ; i++) 
does.
But only those who know this macro know what it hides.

File: tests/posix/env/src/env.c
Comment (aescolar) [inline tests/posix/env/src/env.c:30]: Nit: These macros add a layer of complexity for no apparent benefit. 
The code is actually shorter and more readable without this macro.
```
static DEFINE_ENVIRON(home, "HOME", _m_home);
vs
static char home[] = "HOME =" _m_home;
```

File: tests/posix/env/src/env.c
Comment (aescolar) [inline tests/posix/env/src/env.c:18]: You are already passing a string, STRINGIFY is not needed.

File: tests/posix/env/src/env.c
Comment (aescolar) [inline tests/posix/env/src/env.c:13]: just "1000" would be more readable.

File: subsys/shell/modules/CMakeLists.txt
Comment (cfriedt) [reply subsys/shell/modules/CMakeLists.txt:23]: I'm working on a better solution.

Putting it in the source is fine, but it's several more lines of code than necessary.

File: tests/posix/env/src/env.c
Comment (cfriedt) [reply tests/posix/env/src/env.c:30]: You have already demonstrated that this macro is useful :-)

Thanks for stating your preference, but I will probably keep it.

File: tests/posix/env/CMakeLists.txt
Comment (cfriedt) [reply tests/posix/env/CMakeLists.txt:15]: > Given that this gives trouble until the picolibc fix is merged

Can you clarify? There is zero trouble as far as I am aware.

> and given that this is just to work around checkpatch

Which is fine, since the checkpatch issues are false positives.

>  anyhow this PR triggers other checkpatch checks

Again, false positives (also mentioned in the summary).

> I would either wait for picolibc's fix

Why would that be necessary, specifically? It really isn't.

> have it defined in the source,

Defining it in the source is fine, but for now it creates an unnecessary amount of noise.

> ask for an exception as in this case it is perfectly justified as you want to test setenv() and unsetenv()

Ask for exceptions to support features that are already part of the release plan? That seems like a waste of many people's time.

File: tests/posix/env/testcase.yaml
Comment (cfriedt) [reply tests/posix/env/testcase.yaml:4]: I will probably not be defining in the source for reasons already specified, but in the long run, it should be perfectly fine to do that in source files.

File: tests/posix/env/src/env.c
Comment (cfriedt) [reply tests/posix/env/src/env.c:77]: I'm perfectly fine using this macro in code that I maintain.

I don't really consider this an issue.

File: tests/posix/env/src/env.c
Comment (cfriedt) [reply tests/posix/env/src/env.c:13]: Fixed!

File: tests/posix/env/src/env.c
Comment (cfriedt) [reply tests/posix/env/src/env.c:18]: Fixed!

**Review by ycsin — DISMISSED** [review]
LGTM

**Comment by cfriedt** [comment]
> LGTM

Thanks - I'll rebase to fix merge conflicts 😄

**Comment by cfriedt** [comment]
@ycsin - I also created a sample application (borrowed from your uname sample)

**Review by ycsin — DISMISSED** [review]

File: lib/posix/options/env.c
Comment (jukkar) [inline lib/posix/options/env.c:243]: Looks like a possible memory leak here. If realloc returns null we loose the original content of environ

File: tests/posix/env/src/env.c
Comment (jukkar) [inline tests/posix/env/src/env.c:12]: Wasn't there a rule to avoid using underscore? Do we really need underscore here at all?

Comment (cfriedt) [reply tests/posix/env/src/env.c:12]: there are rules about underscores, bit this doesn't violate any of them. a single underscore followed by a lowercase usually means "only usable within current scope"

Comment (jukkar) [reply tests/posix/env/src/env.c:12]: I see, imho these various underscore rules are really confusing so I try to avoid using them. Anyway, was just wondering about this.

Comment (cfriedt) [reply tests/posix/env/src/env.c]: Fixed!

File: lib/posix/options/env.c
Comment (cfriedt) [reply lib/posix/options/env.c]: Fixed!

**Comment by cfriedt** [comment]
* mitigate possible memory leak by only assigning non-`NULL` return value to `environ` following `realloc()`
* change `_m_home` and other macros to e.g. `M_HOME` in test suite

**Review by ycsin — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**
